File: boj_codes/C++17/1007.cpp
// https://www.acmicpc.net/problem/1007
// 20201130

#include <stdio.h>
#include <limits>
#include <cmath>

int sum_arr(int* arr, int N);
long long combination(int ** grid, long long total_x, long long total_y, int* minus_arr, int index, const int N, long long optimal);

int main(){
    int T;
    scanf("%d", &T);

    for(int i = 0; i < T; i++){
        int N;
        long long total_x = 0, total_y = 0;
        long long optimal = __LONG_LONG_MAX__;
        scanf("%d", &N);

        int** grid = new int*[N];
        int* minus_arr = new int[N];
        for(int j = 0; j < N; j++){
            grid[j] = new int[2];
            scanf("%d %d", &grid[j][0], &grid[j][1]);
            total_x += grid[j][0];
            total_y += grid[j][1];

            minus_arr[j] = 0;
        }

        optimal = combination(grid, total_x, total_y, minus_arr, 0, N, optimal);
        double res = sqrt((double)optimal);

        printf("%lf\n", res);


        for(int j = 0; j < N; j++)
            delete [] grid[j];
        
        delete [] grid;

    }

    return 0;
}

long long combination(int ** grid, long long total_x, long long total_y, int* minus_arr, int index, const int N, long long optimal){
    if(sum_arr(minus_arr, N) == N / 2){
        for(int i = 0; i < N; i ++){
            if(minus_arr[i]){
                total_x -= 2 * grid[i][0];
                total_y -= 2 * grid[i][1];
            }

        }

        long long res = (total_x * total_x) + (total_y * total_y);
        if(res < optimal){
            optimal = res;
        }
        return optimal;
    }

    else if(index == N)
        return optimal;
    
    else{
        minus_arr[index] = 1;
        optimal = combination(grid, total_x, total_y, minus_arr, index + 1, N, optimal);
        minus_arr[index] = 0;
        optimal = combination(grid, total_x, total_y, minus_arr, index + 1, N, optimal);

        return optimal;
    }

}

int sum_arr(int* arr, int N){
    int sum = 0;
    for(int i = 0; i < N; i++)
        sum += arr[i];

    return sum;
}

File: old_contents/boj_codes/Python3/PyPy3/1007.py
# https://www.acmicpc.net/problem/1007
# 20201127

import sys
import math

from itertools import combinations

MAX_VAL = 2e18

if __name__ == "__main__":
    T = int(sys.stdin.readline().rstrip())

    for _ in range(T):
        N = int(sys.stdin.readline().rstrip())
        grid = [list(map(int, sys.stdin.readline().split())) for n in range(N)]

        total_x, total_y = sum(g[0] for g in grid), sum(g[1] for g in grid)

        optimal_len = MAX_VAL

        total_index = list(range(N))
        comb_index = list(combinations(total_index, N // 2))

        for comb in comb_index:
            cur_x = total_x
            cur_y = total_y
            minus_indicies = list(comb)
            for min_idx in minus_indicies:
                cur_x -= 2 * grid[min_idx][0]
                cur_y -= 2 * grid[min_idx][1]

            res = (cur_x ** 2) + (cur_y ** 2)
            if res < optimal_len:
                optimal_len = res


        del grid
        print(math.sqrt(optimal_len))
